File: java/app/src/main/java/com/google/android/gms/samples/wallet/MainActivity.java
/*
 * Copyright 2023 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gms.samples.wallet;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.activity.ComponentActivity;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.IntentSenderRequest;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.gms.pay.PayClient;
import com.google.android.gms.samples.wallet.databinding.ActivitySavePassBinding;
import com.google.android.gms.samples.wallet.viewmodel.WalletViewModel;

public class MainActivity extends ComponentActivity {

    private WalletViewModel model;

    private static final int ADD_TO_GOOGLE_WALLET_REQUEST_CODE = 999;

    private View addToGoogleWalletButtonContainer;
    private View addToGoogleWalletButton;

  /**
   * Initialize the Google Pay API on creation of the activity
   *
   * @see android.app.Activity#onCreate(android.os.Bundle)
   */
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    initializeUi();

    // Check Google Pay availability
    model = new ViewModelProvider(this).get(WalletViewModel.class);

    // Check out Google Wallet availability
    model.canAddPasses.observe(this, this::setAddToGoogleWalletAvailable);
  }

  private void initializeUi() {

    // Use view binding to access the UI elements
    ActivitySavePassBinding layoutBinding = ActivitySavePassBinding.inflate(getLayoutInflater());
    setContentView(layoutBinding.getRoot());

    addToGoogleWalletButton = layoutBinding.addToGoogleWalletButton.getRoot();
    addToGoogleWalletButtonContainer = layoutBinding.passContainer;
    addToGoogleWalletButton.setOnClickListener(v -> {
      addToGoogleWalletButton.setClickable(false);
      model.savePassesJwt(model.genericObjectJwt, this, ADD_TO_GOOGLE_WALLET_REQUEST_CODE);
    });
  }

  /**
   * If getPayApiAvailabilityStatus returned {@code true}, show the "Add to Google Wallet" button.
   * Otherwise, notify the user that Google Wallet is not available. Please adjust to fit in with
   * your current user flow. You are not required to explicitly let the user know if isReadyToPay
   * returns {@code false}.
   *
   * @param available isReadyToPay API response.
   */
  private void setAddToGoogleWalletAvailable(boolean available) {
    if (available) {
      addToGoogleWalletButtonContainer.setVisibility(View.VISIBLE);
    } else {
      Toast.makeText(
          this,
          R.string.google_wallet_status_unavailable,
          Toast.LENGTH_LONG).show();
    }
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
    super.onActivityResult(requestCode, resultCode, data);

    if (requestCode == ADD_TO_GOOGLE_WALLET_REQUEST_CODE) {
      switch (resultCode) {
        case RESULT_OK: {
          Toast
              .makeText(this, getString(R.string.add_google_wallet_success), Toast.LENGTH_LONG)
              .show();
          break;
        }

        case RESULT_CANCELED: {
          // TODO Handle save canceled
          break;
        }

        case PayClient.SavePassesResult.SAVE_ERROR: {
          if (data != null) {
            String apiErrorMessage = data.getStringExtra(PayClient.EXTRA_API_ERROR_MESSAGE);
            // TODO Handle API error
          }
          break;
        }

        default: // TODO Handle other outcomes
      }

      addToGoogleWalletButton.setClickable(true);
    }
  }
}

File: kotlin/app/src/main/java/com/google/android/gms/samples/wallet/MainActivity.kt
/*
 * Copyright 2023 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gms.samples.wallet

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.google.android.gms.samples.wallet.ui.theme.GoogleWalletSampleTheme
import com.google.android.gms.samples.wallet.viewmodel.WalletUiState
import com.google.android.gms.samples.wallet.viewmodel.WalletViewModel
import com.google.wallet.button.WalletButton

class MainActivity : ComponentActivity() {

    private val addToGoogleWalletRequestCode = 1000
    private val model: WalletViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            val payState: WalletUiState by model.walletUiState.collectAsStateWithLifecycle()

            GoogleWalletSampleTheme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background
                ) {
                    Column(
                        modifier = Modifier
                            .padding(20.dp)
                            .fillMaxHeight(),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        when (payState) {
                            is WalletUiState.PassAdded -> AddToGoogleWalletSuccess()
                            is WalletUiState.Available -> AddToGoogleWallet(::requestSavePass)
                            else -> IndeterminateCircularIndicator()
                        }
                    }
                }
            }
        }
    }

    private fun requestSavePass() {
        model.savePassesJwt(model.genericObjectJwt, this, addToGoogleWalletRequestCode)
    }

    @Deprecated("Deprecated and in use by Google Pay")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == addToGoogleWalletRequestCode) {
            when (resultCode) {
                RESULT_OK -> Toast.makeText(
                    this, getString(R.string.add_google_wallet_success), Toast.LENGTH_LONG
                ).show()

                /* Handle other result scenarios
                 * Learn more at: https://developers.google.com/wallet/generic/android#5_add_the_object_to
                 */
                else -> { // Other uncaught errors }
                }
            }
        }
    }
}

@Composable
fun IndeterminateCircularIndicator() = CircularProgressIndicator(
    color = MaterialTheme.colorScheme.surfaceVariant,
    trackColor = MaterialTheme.colorScheme.secondary,
)

@Composable
fun AddToGoogleWallet(onWalletButtonClick: () -> Unit) = Column(
    modifier = Modifier
        .padding(20.dp),
    verticalArrangement = Arrangement.spacedBy(space = 10.dp),
) {
    WalletButton(
        modifier = Modifier.fillMaxWidth(),
        onClick = onWalletButtonClick,
    )
}

@Composable
fun AddToGoogleWalletSuccess() = Column(
    modifier = Modifier
        .padding(20.dp)
        .fillMaxWidth()
        .fillMaxHeight(),
    verticalArrangement = Arrangement.Center,
    horizontalAlignment = Alignment.CenterHorizontally,
) {
    Image(
        contentDescription = null,
        painter = painterResource(R.drawable.check_circle),
        modifier = Modifier
            .width(200.dp)
            .height(200.dp)
    )
    Text(text = stringResource(id = R.string.add_google_wallet_success))
}


@Preview(showBackground = true)
@Composable
fun WalletButtonPreview() {
    WalletButton(onClick = { /*TODO*/ })
}